vector/src/main/java/im/vector/app/features/voicebroadcast/VoiceBroadcastPlayer.kt

```diff
@@ -18,105 +18,225 @@ package im.vector.app.features.voicebroadcast
+    private var voiceBroadcastStateJob: Job? = null
+    private var currentTimeline: Timeline? = null
+        set(value) {
+            field?.removeAllListeners()
+            field?.dispose()
+            field = value
+        }
+
+    private val mediaPlayerListener = MediaPlayerListener()
+    private var timelineListener: TimelineListener? = null
 
     private var currentMediaPlayer: MediaPlayer? = null
-    private var currentPlayingIndex: Int = -1
+    private var nextMediaPlayer: MediaPlayer? = null
+        set(value) {
+            field = value
+            currentMediaPlayer?.setNextMediaPlayer(value)
+        }
+    private var currentSequence: Int? = null
+
     private var playlist = emptyList<MessageAudioEvent>()
-    private val currentVoiceBroadcastEventId
+    private val currentVoiceBroadcastId
         get() = playlist.firstOrNull()?.root?.getRelationContent()?.eventId
 
-    private val mediaPlayerListener = MediaPlayerListener()
-
-    fun play(roomId: String, eventId: String) {
-        val room = session.getRoom(roomId) ?: error("Unknown roomId: $roomId")
+    private var state: State = State.IDLE
+        set(value) {
+            Timber.w("## VoiceBroadcastPlayer state: $field -> $value")
+            field = value
+        }
 
+    fun playOrResume(roomId: String, eventId: String) {
+        val hasChanged = currentVoiceBroadcastId != eventId
         when {
-            currentVoiceBroadcastEventId != eventId -> {
-                stop()
-                updatePlaylist(room, eventId)
-                startPlayback()
-            }
-            playbackTracker.getPlaybackState(eventId) is State.Playing -> pause()
-            else -> resumePlayback()
+            hasChanged -> startPlayback(roomId, eventId)
+            state == State.PAUSED -> resumePlayback()
+            else -> Unit
         }
     }
 
     fun pause() {
         currentMediaPlayer?.pause()
-        currentVoiceBroadcastEventId?.let { playbackTracker.pausePlayback(it) }
+        currentVoiceBroadcastId?.let { playbackTracker.pausePlayback(it) }
+        state = State.PAUSED
     }
 
     fun stop() {
+        // Stop playback
         currentMediaPlayer?.stop()
-        currentMediaPlayer?.release()
-        currentMediaPlayer?.setOnInfoListener(null)
+        currentVoiceBroadcastId?.let { playbackTracker.stopPlayback(it) }
+
+        // Release current player
+        release(currentMediaPlayer)
         currentMediaPlayer = null
-        currentVoiceBroadcastEventId?.let { playbackTracker.stopPlayback(it) }
+
+        // Release next player
+        release(nextMediaPlayer)
+        nextMediaPlayer = null
+
+        // Do not observe anymore voice broadcast state changes
+        voiceBroadcastStateJob?.cancel()
+        voiceBroadcastStateJob = null
+
+        // In case of live broadcast, stop observing new chunks
+        currentTimeline?.dispose()
+        currentTimeline?.removeAllListeners()
+        currentTimeline = null
```

Comment:
I think the two lines above could be removed, since it's done in the `set()` fun of `var currentTimeline` .